### PR TITLE
Determine formatter for 'auto' (no longer fallback to `syntax_tree`).

### PR DESCRIPTION
**Pull request by andyw8**
### Motivation

Ruby LSP supports two formatters, RuboCop and syntax_tree. When formatting is set to `auto`, we check for RuboCop, and use that if available, otherwise we fallback to syntax tree.

This means for the repos which use neither, we were applying syntax tree formatting. This led to unexpected changes in formatting.

### Implementation

A better way is to check for the presence of RuboCop or syntax_tree in `Gemfile.lock`. But we want to avoid false positives through transitive dependencies, so we only check for **direct** dependencies.

We also want to support RuboCop-based gems such as `rubocop-shopify`, so we pattern match on gems beginning named `^rubocop`.

### Automated Tests

Included

### Manual Tests

With `rubyLsp.formatting` set to `auto`:

* Open a repo which doesn't use any formatter, e.g. `ruby/debug`. Add `ruby-lsp` to the Gemfile, pointing to this branch. Ensure that no formatting is applied when saving a file.
* Open a repo which uses RuboCop, e.g. `Shopify/tapioca`. Add `ruby-lsp` to the Gemfile, pointing to this branch. Ensure that RuboCop formatting is applied when saving a file.
* Open a repo which uses Syntax Tree, e.g. https://github.com/andyw8/syntax_tree_example. Add `ruby-lsp` to the Gemfile, pointing to this branch. Ensure that Syntax Tree formatting is applied when saving a file.

With `rubyLsp.formatting` set to `none`:

* Open a repo which uses RuboCop, e.g. Shopify/tapioca. Add ruby-lsp to the Gemfile, pointing to this branch. Ensure that formatting is **not** applied when saving a file.

With `rubyLsp.formatting` set to `rubocop`:

* Open a repo which doesn't use any formatter, e.g. `ruby/debug`. Add `ruby-lsp` to the Gemfile, pointing to this branch. Try format the file and verify you see an error "Formatter is set to 'rubocop' but RuboCop was not found".

With `rubyLsp.formatting` set to `syntax_tree`:

* Open a repo which uses RuboCop, e.g. Shopify/tapioca. Add ruby-lsp to the Gemfile, pointing to this branch. Ensure that **syntax_tree** formatting is applied when saving a file.

(syntax_tree is always available since it ships with Ruby LSP)

**Known Limitations**

On a repo which has syntax_tree but also has a gem named `rubocop-*`, e.g. https://github.com/discourse/discourse it will wrongly use RuboCop formatting.

### CODE REVIEW

File: lib/ruby_lsp/requests/formatting.rb [hunk 1]
```diff
@@ -34,14 +40,30 @@ def initialize(document, formatter: "auto")
         @uri = T.let(document.uri, String)
         @formatter = T.let(
           if formatter == "auto"
-            defined?(Support::RuboCopFormattingRunner) ? "rubocop" : "syntax_tree"
+            if rubocop_as_direct_dependency
```
[the remark this answers is not in the document]
Comment (andyw8) [reply lib/ruby_lsp/requests/formatting.rb:42]: @vinistock would it make sense to persist this on `@store`?

**Comment by andyw8** [comment]
> What do you think about moving this logic to `Executor` instead and just saving the result in the original `@formatter` option?

I've moved it into there. I have a slight concern that `Executor` is growing quite large (> 500 lines), so we might want to starting thinking about if there's anything that could be extracted out of there.

File: lib/ruby_lsp/executor.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@ def initialize(store, message_queue)
       # store
       @store = store
       @message_queue = message_queue
+      @store.formatter = detected_formatter
```
Comment (vinistock) [inline lib/ruby_lsp/executor.rb:15]: When formatter is `auto`, this is going to perform the detection on every single request - including when it's not formatting.

We need to move this to the initialize request, so that we only set it once. [Here](https://github.com/Shopify/ruby-lsp/blob/2b608f98501dd65f6561c33885ed172bbc407ac0/lib/ruby_lsp/executor.rb#L385) where we set the formatter, you can just invoke the logic and set the right one there.

Then only the formatting request will read `@store.formatter`.

File: test/integration_test.rb [hunk 3]
```diff
@@ -385,6 +385,7 @@ def initialize_lsp(enabled_features, experimental_features_enabled: false)
         initializationOptions: {
           enabledFeatures: enabled_features,
           experimentalFeaturesEnabled: experimental_features_enabled,
+          formatter: "rubocop",
```
Comment (andyw8) [inline test/integration_test.rb:388]: (I'm confirming if this is valid).

**Comment by andyw8** [comment]
Discourse is an interesting edge case for this: It uses syntax_tree, but also has a few custom cops in a `rubocop-discourse` gem.

https://github.com/discourse/discourse/blob/552697e194c840b687b00c6125504734121205ef/Gemfile#L164-L171

One thing we could do to help improve the accuracy is to check for a `.streerc` file.

**Comment by vinistock** [comment]
In ambiguous cases, I would say the best option is to configure the formatter instead of using `auto`.

File: test/executor_test.rb [hunk 4]
```diff
@@ -141,4 +141,36 @@ def test_did_close_clears_diagnostics
...
+  end
+
+  def test_gives_rubocop_precedence_if_syntax_tree_also_present
+    stub_dependencies(rubocop: true, syntax_tree: true)
+    RubyLsp::Executor.new(@store, @message_queue).send(
+      :initialize_request,
+      { initializationOptions: { formatter: "auto" } },
+    )
+    assert_equal("rubocop", @store.formatter)
+  end
```
Comment (st0012) [inline test/executor_test.rb:166]: Don't we also want to test if it sets `none` when rubocop and syntax_tree both aren't present?

**Comment by andyw8** [comment]
@st0012 I have verified on Core as suggested.

**Review by st0012 — APPROVED** [review]
Amazing, I've been waiting for this for so long 😂

File: lib/ruby_lsp/executor.rb [hunk 5]
```diff
@@ -260,6 +260,9 @@ def semantic_tokens_full(uri)
 
     sig { params(uri: String).returns(T.nilable(T::Array[Interface::TextEdit])) }
     def formatting(uri)
+      # If formatter is set to `auto` but no supported formatting gem is found, don't attempt to format
+      return [] if @store.formatter == "none"
```
Comment (vinistock) [inline lib/ruby_lsp/executor.rb:264]: We can just avoid the array allocation here.
```suggestion
      return if @store.formatter == "none"
```

**Review by vinistock — APPROVED** [review]
